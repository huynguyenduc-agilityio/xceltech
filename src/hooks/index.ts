export * from './useToast';
export * from './useConfirm';
export * from './useAuth';
export * from './useUser';
export * from './useLeave';
export * from './useAccount';
export * from './usePagination';
export * from './useNotification';
export * from './useJob';
export * from './useDocument';
export * from './useEducation';
export * from './useGuarantor';
export * from './useFamily';
export * from './useFinancial';
