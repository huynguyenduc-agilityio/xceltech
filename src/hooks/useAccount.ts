import { useQuery } from '@tanstack/react-query';

// Constants
import { accountsQueryKeys } from '@/constants';

// Types
import { IInfoUser } from '@/types';

// Services
import { getAccounts } from '@/services';

export const useGetLeaveAccounts = (userId: string) => {
  const { data: leaveAccounts, isFetching: isLeaveAccountsLoading } = useQuery({
    queryKey: [...accountsQueryKeys.list()] as const,
    queryFn: getAccounts,
    select: (data) =>
      data
        ?.filter(({ id }: { id: string }) => id !== userId)
        .map(({ id, firstName, lastName }: IInfoUser) => ({
          value: id,
          label: `${firstName} ${lastName}`,
        })),
  });

  return { leaveAccounts, isLeaveAccountsLoading };
};
