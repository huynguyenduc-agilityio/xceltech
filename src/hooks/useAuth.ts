import { useNavigate, useLocation } from 'react-router-dom';
import { useMutation } from '@tanstack/react-query';

//Types
import { IActiveUser } from '@/types';

// Constants
import {
  ADMIN_PAGE,
  AUTHENTICATION_PAGE,
  RoleAuthentication,
  USER_PAGE,
} from '@/constants';

// Services
import { activateAccount, loginUser, registerUser } from '@/services';

// Stores
import { useUserActions } from '@/stores';

export const useLogin = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { setUser } = useUserActions();

  const {
    mutateAsync: handleLoginUser,
    isPending: isLoginLoading,
    isError: isLoginError,
  } = useMutation({
    mutationFn: loginUser,
    onSuccess: (data) => {
      // Save data in local storage
      if (data) {
        const path = location.pathname;
        const role = data.user.role;
        const isAdmin =
          role === RoleAuthentication.Admin &&
          path === AUTHENTICATION_PAGE.ADMIN_SIGN_IN;

        const isUser =
          [RoleAuthentication.Employee, RoleAuthentication.Candidate].includes(
            role as RoleAuthentication,
          ) && path === AUTHENTICATION_PAGE.USER_SIGN_IN;

        if (isAdmin) {
          setUser(data);
          navigate(ADMIN_PAGE.DASHBOARD);
        }

        if (isUser) {
          setUser(data);
          navigate(USER_PAGE.DASHBOARD);
        }
      }
    },
  });

  return { isLoginLoading, isLoginError, handleLoginUser };
};

export const useRegister = () => {
  const { mutateAsync: handleRegisterUser, isPending: isRegisterLoading } =
    useMutation({
      mutationFn: registerUser,
    });

  return {
    handleRegisterUser,
    isRegisterLoading,
  };
};

export const useActivateAccount = ({ uidb64, token }: IActiveUser) => {
  const { mutateAsync: handleActivateUser, isPending: isActivateLoading } =
    useMutation({
      mutationFn: () => activateAccount({ uidb64, token }),
    });

  return {
    handleActivateUser,
    isActivateLoading,
  };
};
