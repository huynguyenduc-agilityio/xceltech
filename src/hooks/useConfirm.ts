import { useConfirmDialogStore } from '@/stores';

export const useConfirm = () => {
  return useConfirmDialogStore((state) => state.confirm);
};
