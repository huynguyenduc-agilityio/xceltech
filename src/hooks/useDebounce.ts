import { useCallback, useRef } from 'react';

export const useDebounce = <TParam = unknown>(
  callback: (...args: TParam[]) => void,
  delay = 600,
) => {
  const refTime = useRef<NodeJS.Timeout | null>(null);

  const debouncedFn = useCallback(
    (...args: TParam[]) => {
      if (refTime.current) clearTimeout(refTime.current);
      refTime.current = setTimeout(() => {
        callback(...args);
      }, delay);
    },
    [callback, delay],
  );

  return debouncedFn;
};
