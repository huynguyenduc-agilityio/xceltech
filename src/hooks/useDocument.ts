import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

// Constants
import { documentsQueryKeys, MESSAGES } from '@/constants';

// Types
import { MutationType } from '@/types';

// Services
import {
  downloadAllDocuments,
  getDocuments,
  addDocument,
  editDocument,
} from '@/services';

// Utils
import { downloadBlobFile } from '@/utils';

export const useGetDocuments = () => {
  const { data: documents, isFetching: isDocumentsLoading } = useQuery({
    queryKey: documentsQueryKeys.list(),
    queryFn: getDocuments,
  });

  return { documents, isDocumentsLoading };
};

export const useDocumentMutation = ({ type }: { type: MutationType }) => {
  const queryClient = useQueryClient();

  const mutationFn = type === MutationType.Create ? addDocument : editDocument;

  const {
    mutateAsync: handleDocumentMutation,
    isPending: isDocumentMutationLoading,
  } = useMutation({
    mutationFn,
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: documentsQueryKeys.list(),
      });
    },
  });

  return { handleDocumentMutation, isDocumentMutationLoading };
};

export const useDownloadDocuments = () => {
  const { isPending: isLoading, mutateAsync: handleDownloadDocuments } =
    useMutation({
      mutationFn: downloadAllDocuments,
      onSuccess: (data) => {
        if (!data) {
          throw new Error(MESSAGES.COMMON.DOWNLOAD_FAILED);
        }

        downloadBlobFile(data, 'documents.zip');
      },
    });

  return { isLoading, handleDownloadDocuments };
};
