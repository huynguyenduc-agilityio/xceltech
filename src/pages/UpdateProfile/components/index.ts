export { default as JobDetailTab } from './JobDetailTab';
export { default as PersonalDetailTab } from './PersonalDetailTab';
export { default as EducationTab } from './EducationTab';
export { default as GuarantorTab } from './GuarantorTab';
export { default as FamilyTab } from './FamilyTab';
export { default as FinancialTab } from './FinancialTab';
