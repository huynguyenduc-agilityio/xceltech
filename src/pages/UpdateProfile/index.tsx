import { lazy, Suspense } from 'react';

// Constants
import { USER_PAGE } from '@/constants';

// Hooks
import { useGetInfoUser } from '@/hooks';

// Components
import { Breadcrumb, Fallback, SideMenu } from '@/components';

import {
  EducationTab,
  FamilyTab,
  FinancialTab,
  GuarantorTab,
  JobDetailTab,
  PersonalDetailTab,
} from './components';

const ContactForm = lazy(() => import('@/components/Form/ContactForm'));
const NextOfKinForm = lazy(() => import('@/components/Form/NextOfKinForm'));

const BREADCRUMB_ITEMS = [
  { label: 'Dashboard', href: USER_PAGE.DASHBOARD },
  { label: 'Update Profile' },
];

const UpdateProfile = () => {
  const { userInfo } = useGetInfoUser();

  const { name, department, lineManagement, description } = userInfo?.job || {};

  const TAB_LIST = [
    {
      label: 'Personal Details',
      content: <PersonalDetailTab userInfo={userInfo} />,
    },
    {
      label: 'Contact Details',
      content: (
        <Suspense fallback={<Fallback />}>
          <ContactForm initialValues={userInfo} />
        </Suspense>
      ),
    },
    {
      label: 'Next of kin Details',
      content: (
        <Suspense fallback={<Fallback />}>
          <NextOfKinForm initialValues={userInfo?.kin} />
        </Suspense>
      ),
    },
    {
      label: 'Education Qualifications',
      content: <EducationTab />,
    },
    {
      label: 'Guarantor Details',
      content: <GuarantorTab />,
    },
    {
      label: 'Family Details',
      content: <FamilyTab />,
    },
    {
      label: 'Job Details',
      content: (
        <JobDetailTab
          jobInfo={{
            name,
            department,
            lineManagement,
            description,
          }}
        />
      ),
    },
    {
      label: 'Financial Details',
      content: <FinancialTab />,
    },
  ];

  return (
    <div className="flex flex-col gap-11">
      <div className="flex items-center w-full px-[62px] py-8 bg-white">
        <Breadcrumb items={BREADCRUMB_ITEMS} />
      </div>
      <SideMenu tabs={TAB_LIST} />
    </div>
  );
};

export default UpdateProfile;
