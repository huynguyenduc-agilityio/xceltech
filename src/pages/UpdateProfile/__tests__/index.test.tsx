import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { useGetInfoUser } from '@/hooks';

import UpdateProfile from '..';

jest.mock('@/hooks', () => ({
  useGetInfoUser: jest.fn(),
}));

jest.mock('@/components/Form/ContactForm', () => () => <div>ContactForm</div>);

jest.mock('@/components', () => {
  const originalModule = jest.requireActual('@/components');
  return {
    ...originalModule,
    Breadcrumb: () => <nav aria-label="Breadcrumb">Breadcrumb</nav>,
    SideMenu: ({
      tabs,
    }: {
      tabs: {
        label: string;
        content: React.ReactNode;
      }[];
    }) => (
      <div>
        {tabs.map((tab) => (
          <section key={tab.label}>
            <h2>{tab.label}</h2>
            <div>{tab.content}</div>
          </section>
        ))}
      </div>
    ),
    Fallback: () => <div>Loading...</div>,
    NextOfKinForm: () => <div>NextOfKinForm</div>,
  };
});

// Mock profile tab components
jest.mock('../components', () => ({
  EducationTab: () => <div>EducationTab</div>,
  FamilyTab: () => <div>FamilyTab</div>,
  FinancialTab: () => <div>FinancialTab</div>,
  GuarantorTab: () => <div>GuarantorTab</div>,
  JobDetailTab: ({ jobInfo }: { jobInfo: { name?: string } }) => (
    <div>JobDetailTab - {jobInfo?.name}</div>
  ),
  PersonalDetailTab: ({ userInfo }: { userInfo: { name?: string } }) => (
    <div>PersonalDetailTab - {userInfo?.name}</div>
  ),
}));

describe('UpdateProfile', () => {
  const tabLabels = [
    'Personal Details',
    'Contact Details',
    'Next of kin Details',
    'Education Qualifications',
    'Guarantor Details',
    'Family Details',
    'Job Details',
    'Financial Details',
  ];

  const renderAndExpectBase = () => {
    render(<UpdateProfile />);
    expect(
      screen.getByRole('navigation', { name: /breadcrumb/i }),
    ).toBeInTheDocument();
    tabLabels.forEach((label) => {
      expect(screen.getByRole('heading', { name: label })).toBeInTheDocument();
    });
  };

  beforeEach(() => {
    (useGetInfoUser as jest.Mock).mockReturnValue({
      userInfo: {
        name: 'John Doe',
        job: {
          name: 'Developer',
          department: 'Engineering',
          lineManagement: 'Lead',
          description: 'Full-stack developer',
        },
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<UpdateProfile />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render without crashing when userInfo is empty', () => {
    (useGetInfoUser as jest.Mock).mockReturnValue({ userInfo: {} });
    renderAndExpectBase();
    expect(screen.getByText('PersonalDetailTab -')).toBeInTheDocument();
    expect(screen.getByText('JobDetailTab -')).toBeInTheDocument();
  });

  it('should render breadcrumb, tabs and lazy-loaded content', async () => {
    renderAndExpectBase();
    expect(await screen.findByText('ContactForm')).toBeInTheDocument();
    expect(
      screen.getByText('PersonalDetailTab - John Doe'),
    ).toBeInTheDocument();
    expect(screen.getByText('JobDetailTab - Developer')).toBeInTheDocument();
  });

  it('should render all tab labels and content', async () => {
    renderAndExpectBase();
    expect(
      screen.getByText('PersonalDetailTab - John Doe'),
    ).toBeInTheDocument();
    expect(screen.getByText('JobDetailTab - Developer')).toBeInTheDocument();
    expect(await screen.findByText('ContactForm')).toBeInTheDocument();
    expect(screen.getByText('EducationTab')).toBeInTheDocument();
    expect(screen.getByText('FamilyTab')).toBeInTheDocument();
    expect(screen.getByText('FinancialTab')).toBeInTheDocument();
    expect(screen.getByText('GuarantorTab')).toBeInTheDocument();
  });

  it('should render and interact with tabs', async () => {
    render(<UpdateProfile />);
    expect(
      screen.getByText('PersonalDetailTab - John Doe'),
    ).toBeInTheDocument();

    fireEvent.click(screen.getByRole('heading', { name: 'Contact Details' }));
    await waitFor(() => {
      expect(screen.getByText('ContactForm')).toBeInTheDocument();
    });

    fireEvent.click(
      screen.getByRole('heading', { name: 'Education Qualifications' }),
    );
    await waitFor(() => {
      expect(screen.getByText('EducationTab')).toBeInTheDocument();
    });
  });
});
