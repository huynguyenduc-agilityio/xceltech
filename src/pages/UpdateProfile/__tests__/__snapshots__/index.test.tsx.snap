// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProfile should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-11"
  >
    <div
      class="flex items-center w-full px-[62px] py-8 bg-white"
    >
      <nav
        aria-label="Breadcrumb"
      >
        Breadcrumb
      </nav>
    </div>
    <div>
      <section>
        <h2>
          Personal Details
        </h2>
        <div>
          <div>
            PersonalDetailTab - John Doe
          </div>
        </div>
      </section>
      <section>
        <h2>
          Contact Details
        </h2>
        <div>
          <div>
            Loading...
          </div>
        </div>
      </section>
      <section>
        <h2>
          Next of kin Details
        </h2>
        <div>
          <div>
            Loading...
          </div>
        </div>
      </section>
      <section>
        <h2>
          Education Qualifications
        </h2>
        <div>
          <div>
            EducationTab
          </div>
        </div>
      </section>
      <section>
        <h2>
          Guarantor Details
        </h2>
        <div>
          <div>
            GuarantorTab
          </div>
        </div>
      </section>
      <section>
        <h2>
          Family Details
        </h2>
        <div>
          <div>
            FamilyTab
          </div>
        </div>
      </section>
      <section>
        <h2>
          Job Details
        </h2>
        <div>
          <div>
            JobDetailTab - Developer
          </div>
        </div>
      </section>
      <section>
        <h2>
          Financial Details
        </h2>
        <div>
          <div>
            FinancialTab
          </div>
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
